provider/vsphere/vsphere_discover.go

```diff
@@ -0,0 +1,417 @@
+// Package vsphere provides node discovery for VMware vSphere.
+//
+// The package performs discovery by searching vCenter for all nodes matching a
+// certain tag, it then discovers all known IP addresses through VMware tools
+// that are not loopback or auto-configuration addresses.
+//
+// This package requires at least vSphere 6.0 in order to function.
+package vsphere
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/find"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/vic/pkg/vsphere/tags"
+)
+
+// providerLog is the local provider logger. This should be initialized from
+// the provider entry point.
+var logger *log.Logger
+
+// setLog sets the logger.
+func setLog(l *log.Logger) {
+	if l != nil {
+		logger = l
+	} else {
+		logger = log.New(ioutil.Discard, "", 0)
+	}
+}
+
+// discoverErr prints out a friendly error heading for the top-level discovery
+// errors. It should only be used in the Addrs method.
+func discoverErr(format string, a ...interface{}) error {
+	var s string
+	if len(a) > 1 {
+		s = fmt.Sprintf(format, a)
+	} else {
+		s = format
+	}
+	return fmt.Errorf("discover-vsphere: %s", s)
+}
+
+// valueOrEnv provides a way of suppling configuration values through
+// environment variables. Defined values always take priority.
+func valueOrEnv(config map[string]string, key, env string) string {
```

Comment:
Normally in go-discover client libraries support specific environment variables for authentication, etc. Is this added because that is not the case with this client lib?

Reply:
Yeah, the SDK doesn't really give you anything to work with in this regard. I figured I'd as as much as possible to provide options, but we can get rid of the non-connectivity related ones if you want (`VSPHERE_TAG_NAME`, `VSPHERE_TAG_NAME`, `VSPHERE_TIMEOUT`), if you'd rather people be explicit with those options at least.

Reply:
PS: I went ahead and removed these values :slightly_smiling_face: